### ISSUE TITLE
Erasing bytes from the Hex search box leaves blank spaces

### ISSUE THREAD

**Issue by rohitab** [FarGroup/FarManager#607]
### Far Manager version

3.0.6090.0 x64

### OS version

10.0.22623.1095

### Steps to reproduce

1. Press `Alt+F7` in a panel to open _Find file_ dialog
2. Press `Alt+H` to switch to _Hex_ mode
3. Type in some data in the _Containing_ search box, say `01 02 03 04 05 06`
4. Select the first three bytes `01 02 03` and press `Del`, `Backspace` or `Ctrl+X`

### Expected behavior

The _Containing_ search box should display `04 05 06`

### Actual behavior

The _Containing_ search box displays `         04 05 06`, i.e. the three bytes are erased and replaced with spaces. The remaining three bytes remain in the same place, instead of moving left to the beginning of the control. If you now press `Alt+T` and `Alt+H` to switch to _Text_ mode and back to _Hex_ mode, the data is correctly displayed as `04 05 06`, without the leading spaces. Note that regardless of the spaces, the search is performed correctly.

**Cursor Position**
There is also a minor issue with positioning of the cursor in _Hex_ mode. If the previous search was in _Hex_ mode, and you reopen the _Find file_ dialog, and press `Alt+H` to switch focus to the _Containing_ search box, the cursor is positioned at the beginning of the control. If you then press `Alt+T` followed by `Alt+H` to switch to _Text_ mode and back, the cursor is now positioned at the end of the control, which is the correct.

The above issue is also present in the _Search_ dialog of the Viewer. In addition, if the previous search was in _Text_ mode and you reopen _Search_ using `F7`, the cursor is positioned at the end of the _Search for_ box, which is correct. However, if press `Alt+H` to switch to _Hex_ mode, then press `Tab` to switch to the _Search for_ box, the cursor is positioned in the middle of the control (position is equal to length of text). This particular issue is not present in the _Find file_ dialog.

For all unmodified controls, the cursor is normally positioned at the end of the control.

**Copying Hex Data**
If you are in _Hex_ mode of the _Find file_ dialog and the search box contains `01 02 03 04`, and you press `Ctrl+A` followed by `Ctrl+C` to select and copy all text to the clipboard, the clipboard will have:
`01 02 03 04                                                         `
i.e. 68 characters of data. If you repeat the same in Viewer search, the clipboard will contain 383 characters of data. Ideally, the clipboard should only contain 11 characters, i.e. the length of `01 02 03 04`.

**Comment by MKadaner** [comment]
Thank for testing and summarizing all these details. So, basically, Hex Edit control misbehaves in many ways, and slightly differently when on different dialogs. While certainly annoying, all these issues do not seem critical, especially since most, if not all of them have been there for a long time. I may look at Hex Edit later, after I finish a couple of improvements already in the pipeline, but I cannot promise quick turnaround.

**Comment by rohitab** [comment]
No worries. You're right, these issues have been there for a while, and they are definitely not high-priority.

Looking forward to improvements.

**Comment by rohitab** [comment]
I agree with most of what you said. However, instead of using a masked control or a simple plain text control, how about a slightly customized version of the plain text control? All that would be required is to limit the input to hex characters, and automatically format the control on input/paste.

Just to test it out, I wrote a Lua macro that converts the text input in the _Find file_ dialog to a hex input just by using the `DN_EDITCHANGE` message. Most of the issues with the original hex control are not present here.

- You can type in unlimited data
- Copying all the text does not include additional spaces at the end
- Erasing text from the beginning/middle of the control automatically shifts the bytes and does not leave gaps
- You can also paste data into the control, including with or without separators/spaces
- Insert/overwrite modes work as well
- Keys like `Ctrl+BS` also work and erase the complete byte, unlike the existing hex control

Obviously, the code is for demonstration purposes only and is quickly/badly written. There are a few minor bugs, but since this is never going to be used, I didn't see a reason to fix them.

If you want to test it out, please do the following:
1. Extract `Dialog.FindFile.Text2Hex.lua` from the attached archive [Dialog.FindFile.Text2Hex.zip](https://github.com/FarGroup/FarManager/files/10425696/Dialog.FindFile.Text2Hex.zip)
2. Copy `Dialog.FindFile.Text2Hex.lua` to `%FARPROFILE%\Macros\scripts` directory
3. Reload macros using `macro:load` or restart Far
4. Press `Alt+F7` to open the _Find file_ dialog
5. Press `Alt+T` to switch to _Text_ mode
6. Type in data in the _Containing_ box. It should work like a hex input control.

IMPORTANT: Remember to delete `%FARPROFILE%\Macros\scripts\Dialog.FindFile.Text2Hex.lua` otherwise the _Find file_ dialog will not work for searching text.
